src/Rings/MPolyMap/AffineAlgebras.jl

```diff
@@ -0,0 +1,220 @@
+################################################################################
+#
+#  Morphisms between affine algebra homomorphisms
+#
+################################################################################
+
+const AffAlgHom = MPolyAnyMap{DT, CT, Nothing} where {T <: FieldElem,
+                                                       U <: MPolyElem{T},
+                                                       DT <: Union{MPolyRing{T}, MPolyQuo{U}},
+                                                       CT <: Union{MPolyRing{T}, MPolyQuo{U}}}
+
+affine_algebra_morphism_type(::Type{T}) where {T <: Union{MPolyRing, MPolyQuo}} = MPolyAnyMap{T, T, Nothing, elem_type(T)}
+
+affine_algebra_morphism_type(R::T) where {T} = affine_algebra_morphism_type(T)
+
+affine_algebra_morphism_type(::Type{S}, ::Type{T}) where {S <: Union{MPolyRing, MPolyQuo}, T <: Union{MPolyRing, MPolyQuo}} = MPolyAnyMap{S, T, Nothing, elem_type(T)}
+
+affine_algebra_morphism_type(R::S, U::T) where {S, T} = affine_algebra_morphism_type(S, T)
+
+################################################################################
+#
+#  Singular data stuff
+#
+################################################################################
+
+function _singular_ring_domain(f::MPolyAnyMap)
+  return get_attribute!(f, :singular_ring_domain) do
+    singular_ring(domain(f))
+  end
+end
```

Comment:
You can make use of `@attr`:
```suggestion
@attr Any function _singular_ring_domain(f::MPolyAnyMap)
  singular_ring(domain(f))
end
```
You can replace `Any` with the type of `singular_ring(domain(f))` if you know it, this will adjust the return type (by adding a corresponding type assertion), to improve type stability.

Of course the same remark applies to other places you are using `get_attribute!`